runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java

```diff
@@ -201,11 +372,171 @@ public void translateImpulse(
+        RunnerApi.WindowingStrategy windowProto = components.getWindowingStrategiesOrThrow(
+                coll.getWindowingStrategyId());
+        WindowingStrategy<Object, BoundedWindow> windowingStrategy =
+                (WindowingStrategy<Object, BoundedWindow>)
+                        WindowingStrategyTranslation.fromProto(windowProto, rehydratedComponents);
+        windowCoder = WindowedValue.getFullCoder(elementCoder,
+                windowingStrategy.getWindowFn().windowCoder());
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      outputCoders.put(localOutputName, windowCoder);
+      unionCoders.add(windowCoder);
+    }
+    //UnionCoder unionCoder = UnionCoder.of(unionCoders);
+    //TypeInformation<RawUnionValue> typeInformation =
+    //        new CoderTypeInformation<>(unionCoder);
+
+    RunnerApi.ExecutableStagePayload stagePayload;
+    try {
+      stagePayload = RunnerApi.ExecutableStagePayload.parseFrom(transform.getSpec().getPayload());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    String inputPCollectionId =
+            Iterables.getOnlyElement(transform.getInputsMap().values());
+
+    // TODO: is this still relevant?
+    // we assume that the transformation does not change the windowing strategy.
+    RunnerApi.WindowingStrategy windowingStrategyProto =
+            pipeline.getComponents().getWindowingStrategiesOrThrow(
+                    pipeline.getComponents().getPcollectionsOrThrow(
+                            inputPCollectionId).getWindowingStrategyId());
+    final WindowingStrategy<Object, BoundedWindow> windowingStrategy;
+    try {
+      windowingStrategy =
+              (WindowingStrategy<Object, BoundedWindow>) WindowingStrategyTranslation.fromProto(
+                      windowingStrategyProto,
+                      rehydratedComponents);
+    } catch (InvalidProtocolBufferException e) {
+      throw new IllegalStateException(
+              String.format("Unable to hydrate ExecutableStage windowing strategy %s.",
+                      windowingStrategyProto),
+              e);
+    }
+
+    Map<TupleTag<?>, OutputTag<WindowedValue<?>>> tagsToOutputTags = Maps.newHashMap();
+    Map<TupleTag<?>, Coder<WindowedValue<?>>> tagsToCoders = Maps.newHashMap();
+    // TODO: does it matter which output we designate as "main"
+    TupleTag<OutputT> mainOutputTag = new TupleTag(outputs.keySet().iterator().next());
+
+    // TODO: use the id mapping built above
+    // We associate output tags with ids, the Integer is easier to serialize than TupleTag.
+    // The return map of AppliedPTransform.getOutputs() is an ImmutableMap, its implementation is
+    // RegularImmutableMap, its entrySet order is the same with the order of insertion.
+    // So we can use the original AppliedPTransform.getOutputs() to produce deterministic ids.
+    Map<TupleTag<?>, Integer> tagsToIds = Maps.newHashMap();
```

Comment:
This complication is yet another reason to do away with using the old DoFn operator for portable pipelines. It may actually be easier to get things right by copying out the correct bits from there.